internal/webhook/external/dockerhub.go

```diff
@@ -0,0 +1,144 @@
+package external
+
+import (
+	"encoding/json"
+	"fmt"
+	"net/http"
+
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	xhttp "github.com/akuity/kargo/internal/http"
+	"github.com/akuity/kargo/internal/io"
+	"github.com/akuity/kargo/internal/logging"
+)
+
+// dockerHubHandler handles Docker Hub webhook events.
+// It parses the payload to extract the repository name and triggers a refresh
+// for matching Warehouses.
+func dockerHubHandler(
+	c client.Client,
+	namespace string,
+) http.HandlerFunc {
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		ctx := r.Context()
+		logger := logging.LoggerFromContext(ctx).WithValues("path", r.URL.Path)
+		ctx = logging.ContextWithLogger(ctx, logger)
+		logger.Debug("identifying source repository from Docker Hub payload")
+
+		const maxBytes = 2 << 20 // 2MB
+		b, err := io.LimitRead(r.Body, maxBytes)
+		if err != nil {
+			xhttp.WriteErrorJSON(w,
+				xhttp.Error(
+					fmt.Errorf("failed to read request body: %w", err),
+					http.StatusRequestEntityTooLarge,
+				),
+			)
+			return
+		}
+
+		if len(b) == 0 {
+			logger.Debug("empty request body")
+			xhttp.WriteErrorJSON(w,
+				xhttp.Error(
+					fmt.Errorf("empty request body"),
+					http.StatusBadRequest,
+				),
+			)
+			return
+		}
+
+		payload := BuildPayload{}
+
+		if err = json.Unmarshal(b, &payload); err != nil {
```

Comment:
Irrespective of the fact whether you plan to use the library (I mentioned in other comment) or not, I think it would be good to add this logic for parsing into a separate function. This way we can have unit tests for parsing.